Scripts for Fiserv/Smartfolio data

Charlie-Henry opened the pull request:
fiserv_email_pub.py - downloads the emails from S3 and parses out the CSVs and stores them back on S3

fiserv_DB.py - Downloads fiserv records from s3 and changes it to the schema format for storing in a postgres DB
smartfolio_s3.py - Same as above but for smartfolio

parking_socrata.py - Publishes three datasets to socrata: Payments, Transactions, and Processed Payments (Fiserv).

I still need to do some cleanup with the DB schema but that is still up in the air with the passport data being *hopefully* imminent.

Charlie-Henry committed:
Add files via upload

Charlie-Henry committed:
Changes to DB schema

Charlie-Henry committed:
Create readme.md

Charlie-Henry committed:
Email publishing script

Charlie-Henry committed:
Update readme.md

Charlie-Henry committed:
Update readme.md

Charlie-Henry committed:
Fiserv reports to database script

Charlie-Henry committed:
Update readme.md

Charlie-Henry committed:
Smartfolio S3 to DB script

Charlie-Henry committed:
Script that uploads the database to Socrata

Charlie-Henry committed:
Update README.md

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+import ntpath
+from sodapy import Socrata
+from dotenv import load_dotenv
+import os
```
https://www.python.org/dev/peps/pep-0008/#imports

> Imports should be grouped in the following order:
>
> 1. Standard library imports.
> 2. Related third party imports.
> 3. Local application/library specific imports.
> 
> You should put a blank line between each group of imports.

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+
+# Arguments to pick which month to download from S3
+arg_year = 2021
+arg_month = 12
```
create a `main()` func and use CLI args with `argparse` for these inputss

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+        df = pd.read_csv(response.get("Body"))
+        data.append(df)
+
+        print("Loaded CSV File: '%s'" % csv_f)
```
make use of [the logger util](https://github.com/cityofaustin/atd-parking-data/blob/a52354b6d74c0378809b638c355d1e63f880abb8/meters/txn_history.py#L260) instead of printing.

for string formatting, [f-strings](https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings) are the way to go  these days; e.g. this line would be

```python
logger.debug(f"Loaded CSV File: {csv_f}")
```

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+# Renaming columns for schema
+smartfolio = smartfolio.rename(
+    columns={
+        "PAYMENT_MEAN": "payment_method",
```
if possible, it would be good to pull all of these name mappings into a function (at the end?), and store the mappings in a separate file, `models.py`. that will make it easier to maintain and transparent what the input/output fieldnames are

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+        "start_time",
+        "end_time",
+        "amount",
+    ]
```
same here, it would be good to extract this lists of keys into a `models.py` file

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+smartfolio["validated"] = False
+
+# NA duration column must be filled with zeroes
+smartfolio["duration_min"] = smartfolio["duration_min"].fillna(0)
```
So this replaces null/missing/`None` values with `0`? any idea why a duration would be null?

Charlie-Henry replied:
Yes to try to figure out why this may be the case, I ran back 6 months of data and didn't find any rows where this is the case. I ran into this error during development but I can't re-create why and where this was occurring. I'll go ahead and remove this line for now.

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+data = []
+for csv_f in csv_file_list:
+    # Parse the file
+    if ".csv" in csv_f:
```
a would be a little cleaner to do this filtering before you start the `for` loop. 

```python
csv_file_list = [f for f in csv_file_list if f.endswith(".csv")]
```

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+def banking_change(banking_id):
+    # Changes the bank id field to had a leading zero if it is needed
+    if banking_id == "0" or banking_id == "<NA>":
+        return "No Banking ID"
```
the pythonic way to handle values which you will evaluate for truthiness is to make use of `None`. return `None` here instead of "No Banking ID"

johnclary replied:
actually, you can nix this function. see below.

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+## Column wrangling
+smartfolio["id"] = smartfolio["SYSTEM_ID"]
+smartfolio["Banking Id"] = smartfolio["CARD_TRANS_ID"].astype("Int64").astype(str)
+smartfolio["Terminal Code"] = smartfolio["METER_CODE"].astype(str)
```
lets leave bank ID and terminal code as integrers. so nix `astype(str)` in L134 and L135

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+
+    if out[0] == "0":
+        return out[-1 * len(out) + 1 :]
+    return out
```
suggest you refactor/rename this method

```python
def get_invoice_id(banking_id, terminal_code):
    """Create the Inovice ID which is a concatention of the banking ID and device ID

    Args:
        banking_id (int): whatever this is
        terminal_code (int): whatever this is

    Returns:
        int: The formatted invoice ID
    """
    if not banking_id and terminal_code:
        return -1

    # get last 4 digits of terminal code
    terminal_code = str(terminal_code)[-4:]

    # zero-pad bank ID to 6 digits
    banking_id = f"{banking_id:06d}"
    invoice_id = f"{terminal_code}{banking_id}"
    return int(invoice_id)
```

the benefit here is that the entire transform is more transparent. the input/output types are more obvious, as is our concatenation logic.

Charlie-Henry replied:
I had to use pd.isna for checking if banking_id is null, unfortunately it is ambiguous without it. I've also had to add another case where the banking id = 0. I'm not sure what that means, definitely a question for Flowbird. It had to go after the first if statement due to the ambiguity error dealing with "<N/A>" I was getting.

Thanks for the refactor, it looks really great. I've tested it and we're getting the same results.
```
if not banking_id and terminal_code:
        return -1
```

```
if pd.isna(banking_id) and terminal_code:
        return -1

if banking_id == 0:
        return -1

```


https://pandas.pydata.org/pandas-docs/version/1.0.0/user_guide/missing_data.html#na-in-a-boolean-context

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+
+        print("Loaded CSV File: '%s'" % csv_f)
+
+smartfolio = pd.concat(data, ignore_index=True)
```
you've currently got this set up to:
1. Get file list
1. Download all files
2. Merge all data
3. Do transforms
4. Upsert to postgres


I want to suggest that we process each file separately. So,

1. Get file list
2. For each file:
- Do transforms
- Upsert to postgres

This will make it easier to triage if there's an issue (we can log the specific file that failed) and it will require less data to be stored in memory.

Charlie-Henry replied:
Yes makes much more sense,  I've created a `transform` and `to_postgres` function to deal with this on a file-by-file basis.

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+    Returns a list of email files.
+    :return: object
+    """
+    global aws_s3_client
```
rather than using `global`, it's generally better to pass all of the function dependencies as arguments. so

```python
def aws_list_files(year, month, client):
    ...
```

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+
+aws_s3_client = boto3.client(
+    "s3", aws_access_key_id=AWS_ACCESS_ID, aws_secret_access_key=AWS_PASS,
+)
```
this can go in `main()` and get passed around as needed

johnclary commented on meters/smartfolio_s3.py:
```diff
@@ -0,0 +1,218 @@
+# Date/time wrangling
+smartfolio["datetime"] = pd.to_datetime(
+    smartfolio["SERVER_DATE"], format="%Y-%m-%d %H:%M:%S", infer_datetime_format=True
+)
```
i'm learning a lot about pandas! really cool stuff. can you group some of these transforms into smaller functions? the goal is to get to a fairly tidy `main()` function that calls the various transforms

johnclary commented on fiserv/fiserv_email_pub.py:
```diff
@@ -0,0 +1,140 @@
+            object.put(Body=message_decoded)
+
+        # Removes the file from processed folder
+        s3.Object(BUCKET_NAME, email_file).delete()
```
this all looks great! similar to my comments in `smartfolio_s3.py`:

- use a `main()` function
- make use of our logger utility
- rearrange `import` statements

johnclary commented on fiserv/fiserv_DB.py:
```diff
@@ -0,0 +1,165 @@
+import ntpath
+from pypgrest import Postgrest
+from dotenv import load_dotenv
+import os
```
see my comment in `smartfolio_s3.py`

johnclary commented on fiserv/fiserv_DB.py:
```diff
@@ -0,0 +1,165 @@
+)
+
+# Upsert to postgres DB
+two = client.upsert(resource="fiserv_reports_raw", data=payload)
```
Also looks great! Like my comments in `smartfolio_s3.py`
- use a `main()` func and break up transforms logically into smaller functions
- process one file at a time rather than downloading all/uploading once
- move field name lists to a `models.py` file. ideally that can be shared across multiple scripts
- make use of `argparse` and logger utility

johnclary commented on database/docker-entrypoint-initdb.d/init.sql:
```diff
@@ -146,7 +170,7 @@ GRANT ALL ON TABLE api.passport_transactions_raw TO my_api_user;
+GRANT ALL ON TABLE api.flowbird_payments_raw TO my_api_user;
 
 
 --
```
Schema is looking great. One thing to add to all tables is an `updated_at` field, which will be the time the record was updated in our DB. check out [this procedure](https://github.com/cityofaustin/atd-knack-services/blob/47252bfb630110222fd40582af33381cbeac434a/dev/docker-entrypoint-initdb.d/init.sql#L41) and not how we [set it with a trigger](https://github.com/cityofaustin/atd-knack-services/blob/47252bfb630110222fd40582af33381cbeac434a/dev/docker-entrypoint-initdb.d/init.sql#L106) on each column of concern.

And [here](https://github.com/cityofaustin/atd-knack-services/blob/47252bfb630110222fd40582af33381cbeac434a/dev/docker-entrypoint-initdb.d/init.sql#L61) is how you define the column itself:
```sql
 updated_at timestamp with time zone DEFAULT now() NOT NULL
```

johnclary commented on database/parking_socrata.py:
```diff
@@ -0,0 +1,59 @@
+
+
+## Fiserv reports upsert to socrata
+params = {"select": "*", "order": "invoice_id"}
```
see my comments about `updated_at` in `init.sql`.

the idea is that we will use the `updated_at` column as a filter to select the rows we want to upsert to Socrata. that way we don't have to process the entire DB every time we publish.

[here are the postgrest docs](https://postgrest.org/en/v4.1/api.html) on filtering.

johnclary replied:
and you'll need to add `main()` func and a command argument that accepts a date input. we'll use that date in our postgrest query on the `updated_at` column

johnclary commented on database/parking_socrata.py:
```diff
@@ -0,0 +1,59 @@
+
+response = pstgrs.select(resource="flowbird_payments_raw", params=params)
+
+client.upsert(PAYMENTS_DATASET, response)
```
Looks goooood! Like my comments elsewhere:
- rearrange imports
- use a `main()` func
- use `argparse` and logger util
🔥

Charlie-Henry committed:
Payment report to local DB script

I'm currently referencing the "Dev" folder in S3 for now. Will need to change to "Prod" later.

Charlie-Henry committed:
Addressing changes

still need to implement a models.py

Charlie-Henry committed:
Changes to Socrata upload script

Adds logger utils.py

Charlie-Henry committed:
Changes to schema

adds updated_at and passport table.

Charlie-Henry committed:
Adding documentation

Charlie-Henry committed:
Change logger definition

Charlie-Henry committed:
Add files via upload

Charlie-Henry committed:
Fiserv to postgres database script changes

Also adds logger utils

Charlie-Henry committed:
Add files via upload

Charlie-Henry committed:
Update README.md

Charlie-Henry committed:
Adding documentation

Charlie-Henry committed:
Update README.md

Charlie-Henry committed:
Update readme.md

Charlie-Henry commented:
@johnclary thank you for the review. I feel like these are looking much more *production* and now should have implemented all of your comments. 

The one thing I haven't yet is a `models.py` file where I would have all of our transforms/column mapping in one place. We can chat some more about what that would look like if you'd that that implemented before merging. 

Thanks!

Charlie-Henry committed:
Update readme.md

Charlie-Henry committed:
Fix formatting

Charlie-Henry committed:
Add files via upload

Charlie-Henry committed:
Changes to socrata pub to remove match_field

have to drop `match_field` as it contains details

Charlie-Henry committed:
Changes to fiserv to DB script

changes primary key to `batch_number`+`sequence_number`+`invoice_id`

Charlie-Henry committed:
Match field and ID changes to payments script

Changes primary key to `montera_id` which is a unique field for each transaction. Creates a match field for matching back to fiserv data.

Charlie-Henry committed:
script that populates `flowbird_id` in Fiserv DB

Charlie-Henry committed:
updating readme for the new script

Charlie-Henry commented:
I discovered an issue where some `invoice_id` fields were being duplicated. This was likely to continue to get worse. I've new primary keys for both the flowbird reports data and the fiserv reports data. 

A new `match_field` for both datasets was created which contains some of the numbers from the credit cards used which is then concatenated with `invoice_id` to create `match_field`. An additional script is then used (`match_field_processing.py`) to update the Fiserv records with the unique ID of the flowbird data. This is done so the `match_field` can be dropped before posting to Socrata.

@johnclary

johnclary commented:
@Charlie-Henry minor thing—would you update `requirements.txt` with any missing dependencies? looks like it's just pandas and dotenv.

Charlie-Henry committed:
small change to ID data type

Changing from int to string to reflect what the schema and socrata are doing.

Charlie-Henry committed:
Changing requirements

Charlie-Henry committed:
Moving socrata script

Charlie-Henry committed:
Delete utils.py

Charlie-Henry committed:
Moving fiserv-DB script

Charlie-Henry committed:
moving fiserv email pub

Charlie-Henry committed:
moving match field processing

Charlie-Henry committed:
Delete utils.py

Charlie-Henry committed:
Update README.md

Charlie-Henry committed:
Update README.md

Charlie-Henry committed:
Changing to new DB endpoint

Charlie-Henry commented:
@johnclary I changed the endpoint of the database from the local one to the one you set up. I have some additional scripts for flowbird HUB and Passport that we can put in a separate PR. Also I think it would make sense to set up the ETL for the combined parking transaction data which would include both Passport + Flowbird (and no HUB data).
